<div class="alert alert-info">
<font size="5"><b>Комментарий ревьюера</b></font>

Привет Ольга! Меня зовут Марат, и я буду твоим ревьюером. Спешу сообщить что все ключевые этапы в работе выполнены,  с задачей тебе удалось справиться. По поводу обращения - в IT сфере принято общаться на «ты» :) Но, если привычней на «вы», дай знать. Как ревьюера моя задача помочь тебе в развитии, дав хорошие советы. Я внимательно посмотрю твой код, ознакомлюсь с твоими выводами и оставлю комментарии. Где то могу предложить небольшие исправление в коде, но ненавязчиво. Где потребуются уточнения, я оставлю много наводящих вопросов. Они помогут тебя с поиском верного решения.

Все мои комментарии размечены по цветам, для лучшего восприятия. 
    
<div class="alert alert-success">Зеленым цветом и словом «Успех» отмечены особо удачные и элегантные решения, которыми ты можешь гордиться. </div>
        
<div class="alert alert-warning">Желтым и значком словом «Совет», помечены решения у которых есть альтернативные решения, более оптимальные. Ты можешь найти их сразу и доработать проект, или отложить это на потом, для будущих проектах. Проект будет принят и без их доработки. </div>
        
<div class="alert alert-danger"> Красным цветом и значком словом «Ошибка» помечу твои решения, на которые стоит обратить внимание прежде всего. После их доработки проект будет принят. </div>
        
Залог успеха - работа сообща, взаимное уважение и работа в диалоге. Поэтому, помечай свои ответные комментарии на мои реплики заметным цветом или курсивом, так мне будет легче их отслеживать. Пожалуйста, не изменяй и не удаляй мои комментарии. Все это поможет выполнить повторную проверку быстрей.

</div>

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка" data-toc-modified-id="Подготовка-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка</a></span></li><li><span><a href="#Обучение" data-toc-modified-id="Обучение-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обучение</a></span></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Выводы</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

Привет!
Первый раз делаю проект локально, так как в тренажере получаю ошибку в виду умирающего ядра и ошибку no space left даже без BERT!:( Надеюсь что все ок будет локально.

# Проект для «Викишоп»

Интернет-магазин «Викишоп» запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах. То есть клиенты предлагают свои правки и комментируют изменения других. Магазину нужен инструмент, который будет искать токсичные комментарии и отправлять их на модерацию. 

Обучите модель классифицировать комментарии на позитивные и негативные. В вашем распоряжении набор данных с разметкой о токсичности правок.

Постройте модель со значением метрики качества *F1* не меньше 0.75. 

**Инструкция по выполнению проекта**

1. Загрузите и подготовьте данные.
2. Обучите разные модели. 
3. Сделайте выводы.

Для выполнения проекта применять *BERT* необязательно, но вы можете попробовать.

**Описание данных**

Данные находятся в файле `toxic_comments.csv`. Столбец *text* в нём содержит текст комментария, а *toxic* — целевой признак.

**План проекта**




Ознакомиться с данными.

Выполним нормализацию - токенизацию, лемматизацию и переведем текст в вектора

 

Подготовить данные , обучить модели и рассчитать метрики.


Сделать выводы и оформить проект.

## Подготовка

установлю недостающие пакеты

In [1]:
!pip install --upgrade nltk pyspellchecker gensim pandarallel -q

In [2]:
pip install spacy

In [3]:
pip install numpy==1.26.4  pyyaml

Note: you may need to restart the kernel to use updated packages.


Импорт библиотек

In [4]:
from spellchecker import SpellChecker
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from datetime import datetime
import re

import numpy as np
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, f1_score, roc_auc_score

from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report
from nltk.stem import WordNetLemmatizer
import nltk
nltk.download('wordnet')
# будем использовать библиотеку nltk

from nltk.corpus import stopwords
from sklearn.feature_extraction.text import TfidfVectorizer
nltk.download('stopwords')
from nltk.tokenize import RegexpTokenizer
from nltk.stem import WordNetLemmatizer
from nltk.corpus import stopwords

# from spellchecker import SpellChecker
from gensim.models import Word2Vec, FastText
import gensim.downloader as gensim_api

import matplotlib.pyplot as plt
#from wordcloud import WordCloud

from sklearn.model_selection import (cross_val_score, GridSearchCV,
                                     RandomizedSearchCV, train_test_split)
import pandarallel  # Может не работать на Windows
pandarallel.pandarallel.initialize(nb_workers=6, progress_bar=True)
# pandarallel.pandarallel.initialize(progress_bar=True)
from nltk.tokenize import word_tokenize, wordpunct_tokenize
from nltk import sent_tokenize
# sent_tokenize используется для разделения текста на предложения.
nltk.download('punkt')
import spacy
nltk.download('punkt_tab')
pd.set_option('display.max_colwidth', 500)
import warnings
warnings.filterwarnings("ignore")

[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\yar\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\yar\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


INFO: Pandarallel will run on 6 workers.
INFO: Pandarallel will use standard multiprocessing data transfer (pipe) to transfer data between the main process and workers.

https://nalepae.github.io/pandarallel/troubleshooting/


[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\yar\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package punkt_tab to
[nltk_data]     C:\Users\yar\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!



<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:

Собираем все импорты в верхней части, чтобы легче было ориентироваться и добавлять новые по необходимости. 


<div class="alert alert-warning">


Совет:

    
    
- у тебя тут есть лишние импорты, то что ты не использовано или использовано не верно - стоит убрать, чтобы поберечь ресурсы      
  
    

- есть рекомендации PEP-8 при написании кода, в том числе и для импортов. Если интересно можешь почитать [тут](https://pythonworld.ru/osnovy/pep-8-rukovodstvo-po-napisaniyu-koda-na-python.html). Есть что поправить 



In [6]:
#читаем данные
try:
    df = pd.read_csv('https://code.s3.yandex.net/datasets/toxic_comments.csv',index_col=0)
    print('load from current website')
except:
    print('error')


load from current website


<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюера</b></font>

Совет: 


Если не знаешь - чтобы не было столбца  `Unnamed: 0` при чтении файла можно так:


    pd.read_csv(..., index_col=0)

    
(`Unnamed: 0` появляется при не совсем корректном сохранении файла)    


Unnamed: 0 это "след" старых индексов. Если ты уберёшь первые 10 примеров и своего датасета, сохранишь его, а потом откроешь,  то появится столбец Unnamed: 0 начиная с цифры 9, и появится новый индексы начиная с нуля 


Но это мелочь,  даже не нужно ничего исправлять. Просто знай, чтобы увидев такое в чужом коде не удивляться что бы это могло означать

<div class="alert alert-info">
<h2> Комментарий студента v1 <a class="tocSkip"> </h2>

*учтено*
    
</div>

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюераV2</b></font>



Успех 👍:



👍


 

In [7]:
# посмотрим на наши данные
df.head()

text  \
0                                                                                                                                                                                                                                            Explanation\nWhy the edits made under my username Hardcore Metallica Fan were reverted? They weren't vandalisms, just closure on some GAs after I voted at New York Dolls FAC. And please don't remove the template from the talk page since I'm retired now.89.205.38.27   
1                                                                                                                                                                                                                                                                                                                                                                                                     D'aww! He matches this background colour I'm seemingly stuck with. Thanks.  (talk) 21:51, January 11, 2016 (UTC)   
2                                                                                                                                                                                                                                                                            Hey man, I'm really not trying to edit war. It's just that this guy is constantly removing relevant information and talking to me through edits instead of my talk page. He seems to care more about the formatting than the actual info.   
3  "\nMore\nI can't make any real suggestions on improvement - I wondered if the section statistics should be later on, or a subsection of ""types of accidents""  -I think the references may need tidying so that they are all in the exact same format ie date format etc. I can do that later on, if no-one else does first - if you have any preferences for formatting style on references or want to do it yourself please let me know.\n\nThere appears to be a backlog on articles for review so I guess t...   
4                                                                                                                                                                                                                                                                                                                                                                                                                                                  You, sir, are my hero. Any chance you remember what page that's on?   

   toxic  
0      0  
1      0  
2      0  
3      0  
4      0

In [8]:
len(df)

159292

Всего 159292 документа.Проанализируем распределение таргета

In [9]:
df['toxic'].value_counts(normalize=True,dropna=False)

toxic
0    0.898388
1    0.101612
Name: proportion, dtype: float64

дисбаланс есть, учтем это при разбивке данных на тестовую и тренировочную выборки. 

In [10]:
# посмотрим на наши данные
sample = df['text'].loc[19078]
# sample(n=1)
sample

"Dresden's last name is actually Neufeldt, not Gaede. Her father's name is Mark, April's live-in boyfriend. BlondeVikingPrincess November 17, 2006 (UTC)"

выполним очистку данных 

In [11]:
df['text'] = df['text'].str.replace('[^a-zA-Z]', ' ')

In [12]:
sample = df['text'].loc[19078]
sample

"Dresden's last name is actually Neufeldt, not Gaede. Her father's name is Mark, April's live-in boyfriend. BlondeVikingPrincess November 17, 2006 (UTC)"

## Spacy

In [13]:
# проверим на одном документе
nlp = spacy.load('en_core_web_sm')
doc = nlp(sample )
for token in doc:
    print(token.text, token.lemma_)

Dresden Dresden
's 's
last last
name name
is be
actually actually
Neufeldt Neufeldt
, ,
not not
Gaede Gaede
. .
Her her
father father
's 's
name name
is be
Mark Mark
, ,
April April
's 's
live live
- -
in in
boyfriend boyfriend
. .
BlondeVikingPrincess blondevikingprincess
November November
17 17
, ,
2006 2006
( (
UTC UTC
) )


In [14]:
lemma = []

for doc in  nlp.pipe(df['text'].values):
    lemma.append([n.lemma_ for n in doc])

df['text_lemma'] = lemma
df[['text','text_lemma']].head()

text  \
0                                                                                                                                                                                                                                            Explanation\nWhy the edits made under my username Hardcore Metallica Fan were reverted? They weren't vandalisms, just closure on some GAs after I voted at New York Dolls FAC. And please don't remove the template from the talk page since I'm retired now.89.205.38.27   
1                                                                                                                                                                                                                                                                                                                                                                                                     D'aww! He matches this background colour I'm seemingly stuck with. Thanks.  (talk) 21:51, January 11, 2016 (UTC)   
2                                                                                                                                                                                                                                                                            Hey man, I'm really not trying to edit war. It's just that this guy is constantly removing relevant information and talking to me through edits instead of my talk page. He seems to care more about the formatting than the actual info.   
3  "\nMore\nI can't make any real suggestions on improvement - I wondered if the section statistics should be later on, or a subsection of ""types of accidents""  -I think the references may need tidying so that they are all in the exact same format ie date format etc. I can do that later on, if no-one else does first - if you have any preferences for formatting style on references or want to do it yourself please let me know.\n\nThere appears to be a backlog on articles for review so I guess t...   
4                                                                                                                                                                                                                                                                                                                                                                                                                                                  You, sir, are my hero. Any chance you remember what page that's on?   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            text_lemma  
0                                                                                                                                                                                           [Explanation, \n, why, the, edit, make, under, my, username, Hardcore, Metallica, Fan, be, revert, ?, they, be, not, vandalism, ,, just, closure, on, some, gas, after, I, vote, at, New, York, Dolls, FAC, ., and, please, do, not, remove, the, template, from, the, talk, page, since, I, be, retire, now.89.205.38.27]  
1                                                                                                                                                                                                                                                                                                                                                                [D'aww, !, he, match, this, background, colour, I, be, seemingly, stick, with, ., thank, .,  , (, 

<div class="alert alert-info">
<h2> Комментарий студента v1 <a class="tocSkip"> </h2>

*p.s работает долго*
    
</div>

<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюераV2</b></font>



Совет 🤔:


Должно отрабатывать за 20 - 30 минут. Ты использовал неэффективные циклы, стоило попробовать apply 

<div class="alert alert-danger">
<font size="5"><b>Комментарий ревьюера</b></font>

Ошибка:



- WordNetLemmatizer  рабочий вариант, но у него есть особенности, для корректной работы ему нужно передавать не просто слово, но и POS-тег (Part of Speech, части речи). Набираемся ума-разума [тут](https://webdevblog.ru/podhody-lemmatizacii-s-primerami-v-python/) )  Обрати внимание на функцию `get_wordnet_pos`. Сразу хочу предупредить, что если делать Лемматизацию правильно, сучетом постегов, то время может занять полчаса-час. Так что не удивляйся (вообще советую сохранить результаты Лематизации в каком-то файлике, чтобы каждый раз не тратить на это много времени)
 


<div class="alert alert-warning">


Совет: 




- лемматизацию можно было сделать с помощью SpaCy лемматизатором и прямо скажем как инструмент он более удобен и универсален, не нужно заморачиваться с токенизацией и учётом пос тегов




- Чтобы сэкономить время, и убедиться что всё отработало нормально, берёшь парочку предложений, создаёшь dataframe
    
    
    sentence1 = "The striped bats are hanging on their feet for best"
    sentence2 = "you should be ashamed of yourself went worked"
    df_my = pd.DataFrame([sentence1, sentence2], columns = ['text'])
    print(df_my)


    print(df_my['text'].apply(func))
    
    
    
И тестируешь не нем, должно получиться 
    
    
    
    striped  ------> strip, went -------> go  



Если всё получилось, то можно использовать на всём датасете

<div class="alert alert-info">
<h2> Комментарий студента v1 <a class="tocSkip"> </h2>

*заменила лемматизатор на Spacy. При работе над проектом я ориентировалась на вебинар, где был разобран WordNetLemmatizer. На вебинаре к сожалению о тегах упомянуто не было.*
    
</div>

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюераV2</b></font>



Успех 👍:



Понял тебя. Ничего страшного, ревььюера тоже могут добавить полезную информацию) 


 

для проверки орфографии

In [15]:
checker = SpellChecker(distance=1)

In [16]:
df['text_lemma'] = df['text_lemma'].parallel_apply(lambda words:
                   [checker.correction(word) for word in words])
df.head()

NameError: name 'checker' is not defined

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:

 
- Плюс за использование parallel_apply, неэффективные циклы нам ни к чему.


- Да, всегда лучше проверить что получилось  в итоге, так всегда будет возможность поправить ошибку

<div class="alert alert-warning">


Совет: 

 
 
    
- на несбалансированность обратил внимание, это хорошо, но у тебя во всем проекте нет ни одного графика, а ведь красивый, хорошо оформленный график может быть украшением проекта. Почему бы тут его не использовать?   
    
 


- можно также посчитать количество слов в предложений,  длину слов в твите, опять же в разбивке по Таргету.  Если будут какие-то сильные отличия, возможно из-за этого стоит сгенерировать дополнительные признаки для наших моделей. Или например можно использовать библиотеку SentimentIntensityAnalyzer для оценки сантиментов, и посмотреть насколько хорошо ее оценки корелирует с нашими таргетами
   
   
- когда мы работаем с текстами, describe итп описательные статистике не использовать, но можно провести частотный анализ текста.  Предлагаю для этого использовать [облако слов](https://habr.com/ru/post/517410/) - чтобы получить общее представление о тематике и о наиболее часто встречаемых словах в токсичных и нетоксичных твитах (в облаке уже автоматически будут убраны стоп слова). Кроме того графики, рисунки делают проект визуально интересней
   
   
В тренажере облако импортируем так

    !/opt/conda/bin/python -m pip install wordcloud 


или

    !/opt/conda/bin/python -m pip install wordcloud==1.8.2.2  


И возможно дополнительно надо будет сделать



    !pip install --upgrade Pillow  (попробуй версию 9.5.0)

    

разделим данные на тест и трейн, из-за дисбаланса таргета используем stratify

In [17]:

#если классы с дисбалансом - надо применить stratify
X_train, X_test, y_train, y_test = train_test_split(df['text'],
                                                    df['toxic'],
                                                    stratify = df['toxic'],
                                                    test_size=0.25,
                                                    random_state=12345)

print(X_train.shape)
print(X_test.shape)

(119469,)
(39823,)


<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:


- random_state на месте

 

- здорово что используешь stratify    



    
- плюс за  проверку
    
   
    
<div class="alert alert-warning">



Совет 🤔:


Импорты поднимаем наверх, так их легче контролировать 

<div class="alert alert-info">
<h2> Комментарий студента v1 <a class="tocSkip"> </h2>

*импорты перенесены*
    
</div>

## Обучение

### CountVectorizer

**векторизация текста**

Для каждого уникального слова в документе считается количество его вхождений (частота).

Построение векторов: Каждому документу присваивается вектор, длина которого равна количеству уникальных слов во всем корпусе, а значения в векторе — это частоты соответствующих слов в данном документе.

In [18]:
vectorizerCount = CountVectorizer()

In [19]:
x_train_count = vectorizerCount.fit_transform(X_train)

In [20]:
x_train_count.shape

(119469, 159334)

In [21]:
%%time
#Обучаем
regression = LogisticRegression(fit_intercept=True, 
                                class_weight='balanced', 
                                random_state=42,
                                solver='liblinear'
                               )
# словарь 
regression_parametrs = {'C': [0.1, 1, 10]}

# RandomizedSearchCV
regression_grid = RandomizedSearchCV(regression, regression_parametrs, scoring='f1', cv=3)
regression_grid.fit(x_train_count, y_train)

regression.fit(x_train_count, y_train)
regression_cv_score = cross_val_score(regression,x_train_count, y_train,scoring='f1',cv=3).mean()
print('Средняя метрика  логистической регрессии на кросс-валидации:', regression_cv_score)

Среднее качество модели Логистической регрессии на кросс-валидации: 0.7408615805466806
CPU times: total: 3min 9s
Wall time: 3min 1s


<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюераV2</b></font>



Совет 🤔:


Единственное замечание, зачем тебе  cross_val_score, тем более на regression (А это ведь регрессия с дефолтными гиперпараметрами), если ты используешь RandomizedSearchCV? У Неве тоже есть Метрика кросс валидации. Вот так ее смотрим
    
    
    regression_grid.best_score_

In [22]:
%%time

regression_params = regression_grid.best_params_
regression_score = regression_grid.score(x_train_count, y_train)
print(regression_params)
print(regression_score)


{'C': 10}
0.8136286681715575
_______________
CPU times: total: 62.5 ms
Wall time: 66 ms


In [23]:
%%time
# инициализация
forest = RandomForestClassifier(class_weight='balanced', n_jobs=-1 ,random_state=12345 )
# словарь 
forest_parametrs = { 'n_estimators': range(70, 75, 5),
                     'max_depth': range(20, 24, 4),
                     'min_samples_leaf': range(7,9),
                     'min_samples_split': range(4,6,2)}

# RandomizedSearchCV с кросс-валидацией
forest_grid = RandomizedSearchCV(forest, forest_parametrs, scoring='f1', cv=3)
forest_grid.fit(x_train_count, y_train)

CPU times: total: 4min 3s
Wall time: 1min 2s


RandomizedSearchCV(cv=3,
                   estimator=RandomForestClassifier(class_weight='balanced',
                                                    n_jobs=-1,
                                                    random_state=12345),
                   param_distributions={'max_depth': range(20, 24, 4),
                                        'min_samples_leaf': range(7, 9),
                                        'min_samples_split': range(4, 6, 2),
                                        'n_estimators': range(70, 75, 5)},
                   scoring='f1')

In [24]:
%%time
forest_params = forest_grid.best_params_
forest_score = forest_grid.score(x_train_count, y_train)
print(forest_params)
print(forest_score)


{'n_estimators': 70, 'min_samples_split': 4, 'min_samples_leaf': 7, 'max_depth': 20}
0.36582038516733695
_______________
CPU times: total: 4.45 s
Wall time: 1.34 s


In [25]:
%%time
# инициализация
dt = DecisionTreeClassifier(class_weight='balanced',random_state=12345 )
# словарь 
dt_parametrs = { 
                     'max_depth': range(20, 24, 4),
                     'min_samples_leaf': range(7,9),
                     'min_samples_split': range(4,6,2)}

# RandomizedSearchCV с кросс-валидацией
dt_grid = RandomizedSearchCV(dt, dt_parametrs, scoring='f1', cv=3)
dt_grid.fit(x_train_count, y_train)

CPU times: total: 2min 50s
Wall time: 2min 51s


RandomizedSearchCV(cv=3,
                   estimator=DecisionTreeClassifier(class_weight='balanced',
                                                    random_state=12345),
                   param_distributions={'max_depth': range(20, 24, 4),
                                        'min_samples_leaf': range(7, 9),
                                        'min_samples_split': range(4, 6, 2)},
                   scoring='f1')

In [26]:
%%time
dt_params = dt_grid.best_params_
dt_score = dt_grid.score(x_train_count, y_train)
print(dt_params)
print(dt_score)


{'min_samples_split': 4, 'min_samples_leaf': 7, 'max_depth': 20}
0.549044512441545
_______________
CPU times: total: 266 ms
Wall time: 304 ms


<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюераV2</b></font>



Совет 🤔:

Советы по выбору модели для экономии времени (Ведь у нас огромное количество признаков после TFIDF)   


- Деревянные модели (RF, DT) медленные, и на данном датасете не показывает хорошие результаты. Например использование случайного леса часто приводят к получению f1 = 0, потому что для этого датасета нужны деревья бОльше глубины, но это будет слишком долго считать. 

Есть вариант указать в TfidfVectorizer max_features = 7000, и поперебирать n_estimators 100 и больше (плюс возможность взять большую глубину дерева) 


Поэтому стоило выбрать какие-то шустрые модельки, например вот такая [моделька](https://medium.com/geekculture/passive-aggressive-algorithm-for-big-data-models-8cd535ceb2e6) (открывается с помощью VPN) [или](https://datafinder.ru/products/passivno-agressivnyy-klassifikator-v-mashinnom-obuchenii) . Можно с помощью её попробовать получить > 0.75  Она считается очень шустрой 

 

<div class="alert alert-danger">
<font size="5"><b>Комментарий ревьюера</b></font>

Ошибка:


- не забываем при инициализации модели о random_state, иначе после каждого запуска кода у нас может быть разный результат

 
- нет никакого перебора гиперпараметров, а это опять классический подход чтобы поднять метрику     
     
    
    
    
Есть два варианта исправить красное:
    
    
1. Создать валидационный датасет, найти по нему лучшую модель (не по дефолтным значениям, а добавить какой то перебор в цикле).  
    
    
2.  Можно вместо цикла использовать sklearn-ий встроенный функционал GridSearch. В случаи использования GridSearch, не нужно будет заранее делать валидационную выборку, лучшую метрику автоматом сохранят в best_score_, а лучшую модель (переобученная уже на полном наборе данных) будет хранить в best_estimator_, данные обучения положит в cv_restult_. А главное  он сделает несколько разбиений на train / validation выборки (кросс-валидация), тем самым поборется с рандомом, когда на валидации получен хороший результат только изза удачного сплита. 

А еще лучше использовать связку GridSearchCV + pipeline. 

[Pipeline](https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html), это тема которая сразу затрагивает кроссвалидацию, тюнинг "векторайз", подбор гиперпараметров модели и о том что код стоит делать компактным.
    
    
- в TfidfVectorizer(stop_words=stopwords) у тебя по умолчанию ngram_range=(1, 1), тут можно подбирать разное число n- грамм (и другие параметры), максимизируя метрику, но как объединить перебор по ngram_range с обучением моделей, чтобы не делать это по отдельности или с использованием цикла?! pipeline! Готовый [пример для работы с текстами](https://medium.com/@yoni.levine/how-to-grid-search-with-a-pipeline-93147835d916). Всё что нужно там есть, хотя очень лаконично. Можешь погуглить по:


    
    pipeline nlp gridsearchcv
    
    
- как избежать ошибки подглядывания в будущее, когда мы предварительно работаем с данными (шкалирование, нормализация, TfidfVectorizer итп итд)? pipeline! особенно это важно, когда мы используем кроссвалидацию. Для TfidfVectorizer делаеь .fit (обучаемся) на train, а transform на test, точно также нужно сделать для валидационной выборки. Но GS делает валидационные внутри себя, спрашивается как добраться до них и избежать подглядывания в будущее? Казалось бы никак, но нет! Pipeline! ) 
    
    
- pipeline позволяет делать наш код компактней и читабельней, это большой плюс, когда код будет раздуваться   



В общем если сделать GS+pieline будет вообще хорошо )  
    
     

<div class="alert alert-info">
<h2> Комментарий студента v1 <a class="tocSkip"> </h2>

*буду использовать randomized search и кросс валидацию, с рандомайзед все быстрее чем с грид серч*
    
</div>

<div class="alert alert-danger">
<font size="5"><b>Комментарий ревьюера</b></font>

Ошибка:


test датасет используем в самом конце, когда тестируем лучшую модель. Пока не выберем лучшую модель, тестовая выборка как будто бы для нас не существует 





<div class="alert alert-info">
<h2> Комментарий студента v1 <a class="tocSkip"> </h2>

*ОК*
    
</div>

### TF-IDF

In [27]:
vect = TfidfVectorizer(stop_words=stopwords.words('english'),
                       lowercase=True,
                       min_df=0.0001
                       )

features_train = vect.fit_transform(X_train)
print(features_train.shape)

(119469, 17012)


<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюераV2</b></font>



Совет 🤔:


 Кстати используя pipeline, можно было Отдельно не рассматривать TFIDF и CountVectorizer,  а перебирать их, как например мы перебирали гиперпараметр C для логистической регрессии. Так было бы компактнее


#### lr + tfidf

In [28]:
%%time
#обучение
regression_tfidf = LogisticRegression(fit_intercept=True, 
                                class_weight='balanced', 
                                random_state=42,
                                solver='liblinear'
                               )
#словарь
regression_parametrs = {'C': [0.1, 1, 10]}

#RandomizedSearchCV с кросс-валидацией
regression_grid = RandomizedSearchCV(regression_tfidf, regression_parametrs, scoring='f1', cv=3)
regression_grid.fit(features_train, y_train)

regression_tfidf.fit(features_train, y_train)
regression_cv_score = cross_val_score(regression_tfidf,features_train, y_train,scoring='f1',cv=3).mean()
print('Средняя метрика логистической регрессии на кросс-валидации:', regression_cv_score)

Среднее качество модели Логистической регрессии на кросс-валидации: 0.735787260328922
CPU times: total: 12.8 s
Wall time: 12.9 s


In [29]:
%%time

regression_params_tfidf = regression_grid.best_params_
regression_score_tfidf = regression_grid.score(features_train, y_train)
print(regression_params_tfidf)
print(regression_score_tfidf)


{'C': 10}
0.8559948338953863
_______________
CPU times: total: 31.2 ms
Wall time: 51 ms


#### rf + tfidf

In [30]:
%%time
# инициализация
forest = RandomForestClassifier(class_weight='balanced', n_jobs=-1 ,random_state=12345 )
#словарь 
forest_parametrs = { 'n_estimators': range(70, 75, 5),
                     'max_depth': range(20, 24, 4),
                     'min_samples_leaf': range(7,9),
                     'min_samples_split': range(4,6,2)}

# RandomizedSearchCV с кросс-валидацией
forest_grid = RandomizedSearchCV(forest, forest_parametrs, scoring='f1', cv=3)
forest_grid.fit(features_train, y_train)

CPU times: total: 49.7 s
Wall time: 13.6 s


RandomizedSearchCV(cv=3,
                   estimator=RandomForestClassifier(class_weight='balanced',
                                                    n_jobs=-1,
                                                    random_state=12345),
                   param_distributions={'max_depth': range(20, 24, 4),
                                        'min_samples_leaf': range(7, 9),
                                        'min_samples_split': range(4, 6, 2),
                                        'n_estimators': range(70, 75, 5)},
                   scoring='f1')

In [31]:
%%time
forest_params_tfidf = forest_grid.best_params_
forest_score_tfidf = forest_grid.score(features_train, y_train)
print(forest_params_tfidf )
print(forest_score_tfidf )


{'n_estimators': 70, 'min_samples_split': 4, 'min_samples_leaf': 7, 'max_depth': 20}
0.4702013422818792
_______________
CPU times: total: 2.47 s
Wall time: 736 ms


#### dt + tfidf

In [32]:
%%time
# инициализация
dt = DecisionTreeClassifier(class_weight='balanced',random_state=12345 )
#  словарь 
dt_parametrs = { 
                     'max_depth': range(20, 24, 4),
                     'min_samples_leaf': range(7,9),
                     'min_samples_split': range(4,6,2)}

# RandomizedSearchCV с кросс-валидацией
dt_grid = RandomizedSearchCV(dt, dt_parametrs, scoring='f1', cv=3)
dt_grid.fit(features_train, y_train)

CPU times: total: 1min 4s
Wall time: 1min 5s


RandomizedSearchCV(cv=3,
                   estimator=DecisionTreeClassifier(class_weight='balanced',
                                                    random_state=12345),
                   param_distributions={'max_depth': range(20, 24, 4),
                                        'min_samples_leaf': range(7, 9),
                                        'min_samples_split': range(4, 6, 2)},
                   scoring='f1')

In [33]:
%%time
dt_params_tfidf = dt_grid.best_params_
dt_score_tfidf = dt_grid.score(features_train, y_train)
print(dt_params_tfidf)
print(dt_score_tfidf)


{'min_samples_split': 4, 'min_samples_leaf': 7, 'max_depth': 20}
0.6342825237297599
_______________
CPU times: total: 156 ms
Wall time: 167 ms


### Предобученный Word2Vec


попробуем работать с word2vec который уже может работать и с контекстами. Модель обучается на гугл новостях.

TfIdf учитывает только частоты слов, но не учитывает контекста, Word2Vec  учитывает контекст

<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюера</b></font>

Совет: 


Если решила попробовать использовать модели учитывающие контекст,  то предложил бы сразу использовать Берт (word2vec на моей памяти в этом проекте ничего интересного не показывал)


Как мы можем использовать Берт?  



- Можем его использовать чтобы получить эмбединги и подать их в наши модели как альтернативу векторов от TFIDF/CountVectorizer (чтобы это реализовать можно взять готовый код в тренажёре) - Таким образом мы получаем, (если выбрать верную модель)  метрику за 0,9. 
    
    
    
- Можно потюнить модель Берта,  можешь взять на основу [статью (там вообще все возможные варианты рассмотрены, причём с использованием разных библиотек)](https://habr.com/ru/articles/704592/) или этот [ролик](https://www.youtube.com/watch?v=Z1J3sTJYIcc&list=PLEwK9wdS5g0qksxWxtE5c2KuFkIfUXe3i&index=14), там прямо можно посмотреть процесс кодирования, и получше разобраться в практической релизации Берта (тут реализация сложнее, метрики я видел за 0,8). Если использовать [Trainer](https://pytorch.org/rl/reference/generated/torchrl.trainers.Trainer.html), будет всё гораздо проще. 
    
     

<div class="alert alert-info">
<h2> Комментарий студента v1 <a class="tocSkip"> </h2>

*Я тогда лучше просто остановлюсь на тф идф)) Берт - не хватает ресурсов (с коллабом тоже не очень)*
    
</div>

<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюераV2</b></font>



Совет 🤔:


Когда используем Берт, разрешено оставить выборку в 2.000 примеров. И тогда  за полчаса час можно было всё это прогнать 

Таким образом, лучше всех себя проявила логистическая регрессия + tf-idf - мы смотрим на время обучения и метрику. 
выберем ее лучшей моделью и рассчитаем метрику и время на тестовой выборке

In [34]:
# подготовим тестовые данные 
features_test = vect.transform(X_test)

In [35]:
%time
best = LogisticRegression(**regression_params_tfidf)
best.fit(features_train, y_train)





<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюераV2</b></font>
    
    
Совет:



    


Не надо воспринимать  GS как способ получить .best_params_, чтобы подставить их в модель и обучить на них. GS это сделал уже и модельку положил тут: .best_estimator_
    
  
То есть вот это не нужно    
    
    best = LogisticRegression(**regression_params_tfidf)
    best.fit(features_train, y_train)

 

In [ ]:
%time
y_pred = best.predict(features_test)

In [36]:
regression_f1 = round(f1_score(y_test, y_pred), 3) 
print(regression_f1)

0.779


Метрика подходит под условия задания!)

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюераV2</b></font>

Успех: 

- Теперь всё верно, логика моделирования не нарушена 

- Если студент получил на тесте f1 выше 0,75, это считается приемлемым результатом.


<div class="alert alert-warning">



Совет: 

 


А ещё можешь посмотреть какие слова  является наиболее важным для классификации с точки зрения модели. Получаем список слов    
    
    
    
    .get_feature_names_out().tolist()
    
    
    
Получаем коэффициенты важности (для логистической регрессии)    
    
    .coef_.tolist()[0]


 

А потом можно построить такой-то красивый график с помощью     seaborn

 



<div class="alert alert-danger">
<font size="5"><b>Комментарий ревьюера</b></font>

Ошибка:

А вот тут, в самом конце,  выбрав лучшую модель на валидации, проверяем ее на тестовом датасете - делаем финальное тестирование. И если лучшая модель выбранная на валидационной покажет на test результат хуже требуемого, мы начнем процесс моделирования сначала (а не будем такие - "а давай попробуем на тесте модель которая на валидации не была лучшей, может она нам на test даст нужное качество").         
    
Почему только лучшая?! Это делается для того, чтобы мы даже незначительным образом не "подгонялись" под тестовую выборку. Ведь на train модели обучаются, по валидиации подгоняются гиперпараметры. Эти данные модели "знают". А test (out-of-sample) это уже моделирование прогноза на реальных данных и ситуации когда у нас есть уже лучшая модель (в рельности у нас же не может быть несоклько прогнозов, что то в любом случаи надо выбирать). Вот поэтому такая двухуровневая проверка на подгонку. Кроме того использование мноих моделей с разными гиперпараметрами это тоже подгонка, поэтому выбирая одну и тестируя только ее, мы тем самым боремся с подгонкой через использование многих-многих моделей, когда результат хорош не потому что мы данные почистили хорошо, моделировали правильно итд итп, а потому что из многих моделей хоть какая то случайно "сыграет". 




## Выводы

В рамках проекта была проделана большая работа по изучению данных, связанных с работой с текстом. Я применила все знания и умения, полученные в ходе спринта и использовала  информацию из дополнительных источников.

Я открыла файл, посмотрела на кол-во строк, затем текст был разбит токены, примененена лемматизация для приведения слов к начальной форме, проверена орфография  и далее этап векторизации.

Сформированы обучающая и тестовая выборка (0,75% и 0,25%).

Буду работать с тремя моделями: логистическую регрессию, случайный лес и дерево решений.
Сначала я пробовала использовать CountVectorizer. Данная модель отлично подойдет для этого, ведь у нас задача бинарной классификации.

Далее я создала новые признаки с помощью TF-IDF,и повторно обучала  модели. 

Лучшей моделью оказалась модель логистической регрессии с TF-IDF. На кроссвалидации значение метрики оказалось равным 0,86

На тестовых данных также удалось добиться требуемого значения метрики. метрика = 0.779






<div class="alert alert-info">
<font size="5"><b>Комментарий ревьюера</b></font>



Ольга, у тебя старательно выполненная работа, все четко, осмысленно. 



Нет проблем с комментированием кода - всё что ты делаешь понятно


 

Очистка проведена корректно (но есть вопросы к лемматизации). Здорово что попробовала параллельные расчёты


Выводы присутствуют, они четкие и подробные.

 


Как можно сделать проект интереснее, усложнев его:


 

- попробуй использовать связку GS+pipeline




- попробуй очень современный и модный сейчас подход с использованием  Берта. Есть несколько вариантов, самый эффективный - это использовать эмбединги (как замена TFIDF) для этого есть уже готовый код в тренажёре (в этом случае разрешается сильно порезать датасет, а если еще и использовать GPU в Colab код можно прогнать за полчаса).



- для красивой  визуализации можешь построить облако для токсичных и нетоксичных комментариев



- чтобы заглянуть внутрь модели, можешь посмотреть какой из признаков является наиболее важный для логистической регресси (должно получиться слово fuck)   

 
Обязательное к исправлению:



 


- WordNetLemmatizer используем с POS - тег 
 


- при инициализации модели не забываем random_state (можно один раз вначале просто прописать random.seed(42), чтобы не прописывать каждый раз везде random_state. Кстати знаешь откуда 42?)



 



- нет перебора гиперпараметров (это можно сделать с помощью вручную прописанных циклов либо с помощью GridSearchCV)
 


- на test датасете тестируем только лучшую модель (нарушена логика использования датасетов при моделировании)


 


Жду исправлений, для принятия проекта. Если какие то вопросы, то сразу спрашивай ) 


<font color='green'><b>Полезные (и просто интересные) материалы:</b> \
Для работы с текстами используют и другие подходы. Например, сейчас активно используются RNN (LSTM) и трансформеры (BERT и другие с улицы Сезам, например, ELMO). НО! Они не являются панацеей, не всегда они нужны, так как и TF-IDF или Word2Vec + модели из классического ML тоже могут справляться. \
BERT тяжелый, существует много его вариаций для разных задач, есть готовые модели, есть надстройки над библиотекой transformers. Если, обучать BERT на GPU (можно в Google Colab или Kaggle), то должно быть побыстрее.\
https://huggingface.co/transformers/model_doc/bert.html \
https://t.me/renat_alimbekov \
https://colah.github.io/posts/2015-08-Understanding-LSTMs/ - Про LSTM \
https://web.stanford.edu/~jurafsky/slp3/10.pdf - про энкодер-декодер модели, этеншены\
https://pytorch.org/tutorials/beginner/transformer_tutorial.html - официальный гайд
по трансформеру от создателей pytorch\
https://transformer.huggingface.co/ - поболтать с трансформером \
Библиотеки: allennlp, fairseq, transformers, tensorflow-text — множествореализованных
методов для трансформеров методов NLP \
Word2Vec https://radimrehurek.com/gensim/models/word2vec.html 


Если понравилась работа с текстами, то можешь посмотреть очень интересный (но очень-очень сложный) курс лекций: https://github.com/yandexdataschool/nlp_course .

Если нравится смотреть и слушать то есть целый курс на Ютубе https://www.youtube.com/watch?v=qDMwIQRQt-M&list=PLEwK9wdS5g0qksxWxtE5c2KuFkIfUXe3i&index=1


<div class="alert alert-info">
<h2> Комментарий студента v1 <a class="tocSkip"> </h2>


    

Обязательное к исправлению:



 


- WordNetLemmatizer используем с POS - тег - **Теперь я использую spacy**
 


- при инициализации модели не забываем random_state (можно один раз вначале просто прописать random.seed(42), чтобы не прописывать каждый раз везде random_state. Кстати знаешь откуда 42?) - **RandomState использую. откуда 42 не знаю)**



 



- нет перебора гиперпараметров (это можно сделать с помощью вручную прописанных циклов либо с помощью GridSearchCV) - **Реализовано с использованием randomized search cv**
 


- на test датасете тестируем только лучшую модель (нарушена логика использования датасетов при моделировании) - **исправлено**

</div>


 

<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюерV2</b></font>

Совет: 


По поводу 42
    
https://ru.wikipedia.org/wiki/%D0%9E%D1%82%D0%B2%D0%B5%D1%82_%D0%BD%D0%B0_%D0%B3%D0%BB%D0%B0%D0%B2%D0%BD%D1%8B%D0%B9_%D0%B2%D0%BE%D0%BF%D1%80%D0%BE%D1%81_%D0%B6%D0%B8%D0%B7%D0%BD%D0%B8,_%D0%B2%D1%81%D0%B5%D0%BB%D0%B5%D0%BD%D0%BD%D0%BE%D0%B9_%D0%B8_%D0%B2%D1%81%D0%B5%D0%B3%D0%BE_%D1%82%D0%B0%D0%BA%D0%BE%D0%B3%D0%BE



 


<div class="alert alert-info">
<font size="5"><b>Комментарий ревьюераV2</b></font>

Спасибо за работу!    

    
Красное исправлено. Надеюсь мои советы и вопросики были полезны и в копилочку знаний упало что то новое, а проект стал лучше, и симпатичней. Также дополнительно написал О том что стоило использовать apply, best_score_, best_estimator 
    
    
Красного нет, вопросов нет, значит все, пора принимать) Надеюсь мои советы и вопросики были полезны и в копилочку знаний упало что то новое, а проект стал лучше, и симпатичней.

  
Отличная работа Ольга. Желаю успехов в дальнейшей учебе!

 

## Чек-лист проверки

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Данные загружены и подготовлены
- [x]  Модели обучены
- [x]  Значение метрики *F1* не меньше 0.75
- [x]  Выводы написаны